pkg/client/cli/cmd.go

```diff
@@ -32,6 +32,30 @@ Unless the daemons are already started, an attempt will be made to start them.
 This will involve a call to sudo unless this command is run as root (not
 recommended) which in turn may result in a password prompt.`
 
+type calledSubcommandNameKey struct{}
+
+func WithCalledSubcommandName(ctx context.Context) context.Context {
+	var x string
+	return context.WithValue(ctx, calledSubcommandNameKey{}, &x)
+}
+
+func setCalledSubcommandName(ctx context.Context, name string) {
+	n, _ := ctx.Value(calledSubcommandNameKey{}).(*string)
+	if n == nil {
+		return
+	}
+	*n = name
+}
+
+func GetCalledSubcommandName(ctx context.Context) string {
+	n, _ := ctx.Value(calledSubcommandNameKey{}).(*string)
+	if n == nil {
+		return ""
+	}
+
+	return *n
+}
+
```

Comment:
While this will surely work, it bugs me a little. A cleaner approach would be to move the error handling from main into a PersistentPostRunE. That way, I believe the sub command will be passed as an argument and the `PersistentPreRun` and context variable that you add wouldn't be necessary.

Reply:
> While this will surely work, it bugs me a little. A cleaner approach would be to move the error handling from main into a `PersistentPostRunE`. That way, I believe the sub command will be passed as an argument and the `PersistentPreRun` and context variable that you add wouldn't be necessary.

After looking into this, there's no way to get the error from `RunE` into `PersistentPostRunE` for processing through the mechanisms provided by cobra. We'd have to pass the error through the context, through a root command struct, etc.

Reply:
Could we call it directly from each of the subcommands then? So a CLI command like `cmd_list` would do this?
```
		RunE: func(cmd *cobra.Command, args []string) error {
			return HandleError(cmd, s.list(cmd, args))
		},
```
Then we could also relieve the daemons from this error handling. It doesn't make much sense there anyway.

It is kind of weird that there's no way to handle sub-command errors generically. Are we overlooking something?